pkg/kubelet/pod_workers.go

```diff
@@ -283,8 +296,34 @@ func (p *podWorkers) checkForUpdates(uid types.UID) {
 		p.podUpdates[uid] <- workUpdate
 		delete(p.lastUndeliveredWorkUpdate, uid)
 	} else {
-		p.isWorking[uid] = false
+		p.isWorking[uid] = nil
+	}
+}
+
+// isShortenedGracePeriod returns true if new deletion time is earlier than the original
+func isShortenedGracePeriod(old, new *UpdatePodOptions) bool {
+	need := (old.Pod.DeletionTimestamp != nil || old.UpdateType == kubetypes.SyncPodKill) &&
+		(new.Pod.DeletionTimestamp != nil || new.UpdateType == kubetypes.SyncPodKill)
+	if !need {
+		return false
+	}
+	now := time.Now()
+	oldDeletionTimestamp := getDeleteTimestamp(old, now)
```

Comment:
/hold

No component should EVER read deletion timestamp and act on it.

You may *only* act on the current time + deletionSeconds compared to when you saw the event the first time.

Reply:
The logic of deleting a pod or shortening grace period should be:

1. When the kubelet sees a pod is deleted for the first time, it should take the current time (`time.Now()`) and record that as the time the first deletion was seen `t_delete`, and then trigger a container TERM (stopContainer) with an expected time to force terminate `t_kill = t_delete + max(grace_period, 3s)`.
2. pod stop should only be invoked one time on a pod as long as the kubelet has not restarted
3. If grace period is reduced after stopContainer is requested at least once, we must reduce the time subtracted from the original request (original grace period - current grace period) and execute the kill (SIGKILL, not TERM) to the container runtime, but make sure than no less than 3s have elapsed - so the updated `t_kill = t_delete + max(new_grace_period, 3s)`.  If `t_kill > t_now`, we should kill the pod immediately.

Reply:
But it's really important that shortening deletion time should NEVER cause a second SIGTERM to be sent - it's the container runtimes responsibility to only send SIGTERM to a terminated process once, and then SIGKILL after the elapsed timeout has occurred.

Reply:
I would expect the runtime to not dispatch TERM twice if it had already dispatched it once prior.

I would expect the runtime to just adjust its waiting period before sending SIGKILL.

This is why we need runtimes to ensure they do the right thing as I prefer kubelet stays simple.

/cc @mrunalp

Reply:
Yeah, we are working on a patch for this and should be able to ensure that only one SIGTERM is sent.

Reply:
Containerd will also need a patch to mirror the (new?) desire for only sending one SIGTERM when two CRI container stop requests are received for the same container. Concerned about possibility of regressions here.